### PR TITLE
Add scala documentation

### PR DESCRIPTION
**Pull request by psyyz10**
## What changes were proposed in this pull request?

Add scala documentation

### CODE REVIEW

File: spark/dl/src/main/scala/com/intel/analytics/bigdl/nn/abstractnn/AbstractCriterion.scala [hunk 1]
```diff
@@ -38,20 +58,54 @@ abstract class AbstractCriterion[A <: Activity: ClassTag, B <: Activity: ClassTa
...
+   */
   def backward(input: A, target: B): A = {
     updateGradInput(input, target)
   }
 
+  /**
+   * Computes the loss using input and objective function. This function
+   * returns the result which is stored in the output field.
+   *
+   * @param input input of the criterion
```
Comment (qiuxin2012) [inline spark/dl/src/main/scala/com/intel/analytics/bigdl/nn/abstractnn/AbstractCriterion.scala]: param target is missing

File: spark/dl/src/main/scala/com/intel/analytics/bigdl/optim/Optimizer.scala [hunk 2]
```diff
@@ -54,8 +65,16 @@ abstract class Optimizer[T: ClassTag, D](
   protected var computeThresholdbatchSize: Int = 100
   protected var warmupIterationNum: Int = 200
 
+  /**
+   * Trigger the opmize process
```
Comment (qiuxin2012) [inline spark/dl/src/main/scala/com/intel/analytics/bigdl/optim/Optimizer.scala]: type error here.

File: spark/dl/src/main/scala/com/intel/analytics/bigdl/optim/Optimizer.scala [hunk 3]
```diff
@@ -54,8 +65,16 @@ abstract class Optimizer[T: ClassTag, D](
...
+   * @return the model to be trained
+   */
   def optimize(): Module[T]
 
+  /**
+   * make optimzer not check the singleton model on a node
```
Comment (qiuxin2012) [inline spark/dl/src/main/scala/com/intel/analytics/bigdl/optim/Optimizer.scala]: type error

File: spark/dl/src/main/scala/com/intel/analytics/bigdl/optim/Optimizer.scala [hunk 4]
```diff
@@ -67,6 +86,14 @@ abstract class Optimizer[T: ClassTag, D](
   protected var checkSingleton = System.getProperty("bigdl.check.singleton",
     true.toString).toBoolean
 
+  /**
+   * Set a validate evaluation
+   *
+   * @param trigger how offten to evaluation validation set
```
Comment (qiuxin2012) [inline spark/dl/src/main/scala/com/intel/analytics/bigdl/optim/Optimizer.scala]: type error here.

File: spark/dl/src/main/scala/com/intel/analytics/bigdl/optim/Optimizer.scala [hunk 5]
```diff
@@ -76,6 +103,15 @@ abstract class Optimizer[T: ClassTag, D](
     this
   }
 
+  /**
+   * Set a validate evaluation
+   *
+   * @param trigger how offten to evaluation validation set
```
Comment (qiuxin2012) [inline spark/dl/src/main/scala/com/intel/analytics/bigdl/optim/Optimizer.scala]: type error here.

File: spark/dl/src/main/scala/com/intel/analytics/bigdl/optim/SGD.scala [hunk 6]
```diff
@@ -129,6 +129,14 @@ object SGD {
     def updateHyperParameter(config : Table, state : Table) : Unit
   }
 
+  /**
+   * [[EpochSchedule]] is a learning rate schedule which configure the learning
+   * rate according to some pre-defined [[Regime]]. If the running epoch is within
+   * the interval of a regime `r` [r.startEpoch, r.endEpoch], then the learning
+   * rate will take the "learningRate" in r.config.
```
Comment (qiuxin2012) [inline spark/dl/src/main/scala/com/intel/analytics/bigdl/optim/SGD.scala]: r.config may define a few hyperParameters, not only the LR

File: spark/dl/src/main/scala/com/intel/analytics/bigdl/optim/SGD.scala [hunk 7]
```diff
@@ -170,6 +178,14 @@ object SGD {
     }
   }
 
+  /**
+   * It is an epoch decay learning rate schedule
+   * The learning rate decays through a function argument on number of run epochs
+   *
+   * l_{n + 1} = l_{n} * 0.1 `^` decayType(epoch)
```
Comment (qiuxin2012) [inline spark/dl/src/main/scala/com/intel/analytics/bigdl/optim/SGD.scala]: what does n mean?

File: spark/dl/src/main/scala/com/intel/analytics/bigdl/optim/SGD.scala [hunk 8]
```diff
@@ -195,6 +218,15 @@ object SGD {
     }
   }
 
+  /**
+   * It is the default learning rate schedule.
+   * For each iteration, the learning rate would
+   * update with the following formula:
+   *
+   * l_{n + 1} = l / (1 + n * learning_rate_decay)
```
Comment (qiuxin2012) [inline spark/dl/src/main/scala/com/intel/analytics/bigdl/optim/SGD.scala]: what does n mean?

File: spark/dl/src/main/scala/com/intel/analytics/bigdl/optim/SGD.scala (hunk 6 above)
Comment (psyyz10) [inline spark/dl/src/main/scala/com/intel/analytics/bigdl/optim/SGD.scala]: Yes, but doesn't the new learning rate take the learning rate in the r.config?

File: spark/dl/src/main/scala/com/intel/analytics/bigdl/optim/Optimizer.scala (hunk 5 above)
Comment (psyyz10) [inline spark/dl/src/main/scala/com/intel/analytics/bigdl/optim/Optimizer.scala]: Done

File: spark/dl/src/main/scala/com/intel/analytics/bigdl/optim/Optimizer.scala (hunk 4 above)
Comment (psyyz10) [inline spark/dl/src/main/scala/com/intel/analytics/bigdl/optim/Optimizer.scala]: Done

File: spark/dl/src/main/scala/com/intel/analytics/bigdl/nn/abstractnn/AbstractCriterion.scala (hunk 1 above)
Comment (psyyz10) [inline spark/dl/src/main/scala/com/intel/analytics/bigdl/nn/abstractnn/AbstractCriterion.scala]: Done

File: spark/dl/src/main/scala/com/intel/analytics/bigdl/optim/Optimizer.scala (hunk 3 above)
Comment (psyyz10) [inline spark/dl/src/main/scala/com/intel/analytics/bigdl/optim/Optimizer.scala]: Done

File: spark/dl/src/main/scala/com/intel/analytics/bigdl/optim/SGD.scala (hunk 7 above)
Comment (psyyz10) [inline spark/dl/src/main/scala/com/intel/analytics/bigdl/optim/SGD.scala]: n means how many time the learning rate updates

File: spark/dl/src/main/scala/com/intel/analytics/bigdl/optim/SGD.scala (hunk 8 above)
Comment (psyyz10) [inline spark/dl/src/main/scala/com/intel/analytics/bigdl/optim/SGD.scala]: n means how many time the learning rate updates
